src/planner/vectorfield/detail/VectorFieldIntegrator.cpp

```diff
@@ -0,0 +1,172 @@
+
+    stateSpace->expMap(currentPosition, currState);
+    outputTrajectory->addSegment(coefficients, segmentDuration, currState);
+  }
+  return outputTrajectory;
+}
+
+//==============================================================================
+VectorFieldIntegrator::VectorFieldIntegrator(
+    const VectorField* vectorField,
+    const aikido::constraint::Testable* collisionFreeConstraint,
+    double timelimit,
+    double checkConstraintResolution)
+  : mVectorField(vectorField)
+  , mConstraint(collisionFreeConstraint)
+  , mTimelimit(timelimit)
+  , mConstraintCheckResolution(checkConstraintResolution)
+{
+  mCacheIndex = -1;
+  mIndex = 0;
+
+  mDimension = mVectorField->getStateSpace()->getDimension();
+}
+
+//==============================================================================
+void VectorFieldIntegrator::start()
+{
+  mTimer.start();
+  mKnots.clear();
+  mCacheIndex = -1;
+  mIndex = 0;
+}
+
+//==============================================================================
+int VectorFieldIntegrator::getCacheIndex()
+{
+  return mCacheIndex;
+}
+
+//==============================================================================
+std::vector<Knot>& VectorFieldIntegrator::getKnots()
+{
+  return mKnots;
+}
+
+//==============================================================================
+void VectorFieldIntegrator::step(
+    const Eigen::VectorXd& q, Eigen::VectorXd& qd, double /*_t*/)
+{
+  aikido::statespace::StateSpace::State* state
+      = mVectorField->getStateSpace()->allocateState();
+  mVectorField->getStateSpace()->expMap(q, state);
+
+  // compute joint velocities
+  bool success = mVectorField->evaluateVelocity(state, qd);
+  mVectorField->getStateSpace()->freeState(state);
+  if (!success)
+  {
+    throw IntegrationFailedError();
+  }
+}
+
+//==============================================================================
+void VectorFieldIntegrator::check(const Eigen::VectorXd& q, double t)
+{
+  if (mTimer.getElapsedTime() > mTimelimit)
+  {
+    // throw TimeLimitError();
+  }
+
+  aikido::statespace::StateSpace::State* state
+      = mVectorField->getStateSpace()->allocateState();
+  mVectorField->getStateSpace()->expMap(q, state);
+
+  Knot knot;
+  knot.mT = t;
+  knot.mPositions = q;
+
+  double trajEvalStartTime = 0;
+  if (mKnots.size() > 0)
+  {
+    // create new incremental segment
+    std::vector<Knot> segment;
+    segment.push_back(mKnots.back());
+    segment.push_back(knot);
+    auto trajSegment = convertToSpline(segment, mVectorField->getStateSpace());
+
+    if (mKnots.size() > 1)
+    {
+      trajEvalStartTime = mConstraintCheckResolution;
+    }
+    if (!mVectorField->evaluateTrajectory(
+            *trajSegment.get(),
+            mConstraint,
+            mConstraintCheckResolution,
+            trajEvalStartTime))
+    {
+      throw VectorFieldTerminated("State not in constraint.");
+    }
+  }
+  mKnots.push_back(knot);
+  mIndex += 1;
```

Comment:
It seems like `mIndex` always equals `mKnots.size()`. Is the variable necessary? 🤔